From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: FreshTomato
Date: Tue, 21 Jul 2026 00:00:00 +0200
Subject: [PATCH] uClibc: make byte-swap macros usable from C++ constexpr code

Old uClibc byteswap macros declare temporary result variables without an
initializer:

  unsigned int __bsv, __bsx = (x);

GCC 7 diagnoses this when the macro is expanded in a C++ constexpr
function, even though all runtime branches assign the temporary.  This
is exposed by Transmission's IPv6 predicates through htonl()/htons().

Use GCC byte-swap builtins where available.  Besides avoiding the
uninitialized temporary, GCC can constant-fold the builtins and emit the
native ARM REV/REV16 instruction for non-constant values.

This is a small backport of the approach adopted by uClibc-ng when it
updated bits/byteswap-common.h from glibc.

---
 libc/sysdeps/linux/common/bits/byteswap-common.h | 47 ++++++++++++++----------
 1 file changed, 27 insertions(+), 20 deletions(-)

diff --git a/libc/sysdeps/linux/common/bits/byteswap-common.h b/libc/sysdeps/linux/common/bits/byteswap-common.h
--- a/libc/sysdeps/linux/common/bits/byteswap-common.h
+++ b/libc/sysdeps/linux/common/bits/byteswap-common.h
@@ -32,14 +32,15 @@
 # define __bswap_non_constant_16(x) __bswap_constant_16(x)
 #endif
 #ifdef __GNUC__
-# define __bswap_16(x) \
-    (__extension__							      \
-     ({ unsigned short int __bsv, __bsx = (x);				      \
-	if (__builtin_constant_p (__bsx))				      \
-	  __bsv = __bswap_constant_16 (__bsx);				      \
-	else								      \
-	  __bsv = __bswap_non_constant_16 (__bsx);			      \
-	__bsv; }))
+# if (__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 8))
+#  define __bswap_16(x) \
+     __builtin_bswap16 ((unsigned short int) (x))
+# else
+#  define __bswap_16(x) \
+    (__extension__							      \
+     ({ unsigned short int __bsx = (unsigned short int) (x);		      \
+	__bswap_constant_16 (__bsx); }))
+# endif
 #else
 static __inline unsigned short int
 __bswap_16 (unsigned short int __bsx)
@@ -57,14 +58,15 @@ __bswap_16 (unsigned short int __bsx)
 # define __bswap_non_constant_32(x) __bswap_constant_32(x)
 #endif
 #ifdef __GNUC__
-# define __bswap_32(x) \
-    (__extension__							      \
-     ({ unsigned int __bsv, __bsx = (x);				      \
-	if (__builtin_constant_p (__bsx))				      \
-	  __bsv = __bswap_constant_32 (__bsx);				      \
-	else								      \
-	  __bsv = __bswap_non_constant_32 (__bsx);			      \
-	__bsv; }))
+# if (__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 3))
+#  define __bswap_32(x) \
+     __builtin_bswap32 ((unsigned int) (x))
+# else
+#  define __bswap_32(x) \
+    (__extension__							      \
+     ({ unsigned int __bsx = (unsigned int) (x);			      \
+	__bswap_constant_32 (__bsx); }))
+# endif
 #else
 static __inline unsigned int
 __bswap_32 (unsigned int __bsx)
@@ -95,14 +97,16 @@ __bswap_32 (unsigned int __bsx)
 	 __r.__l[1] = __bswap_non_constant_32 (__w.__l[0]);		      \
 	 __r.__ll; }))
 # endif
-# define __bswap_64(x) \
-     (__extension__							      \
-      ({ __extension__ unsigned long long int __ll;			      \
-         if (__builtin_constant_p (x))					      \
-	   __ll = __bswap_constant_64 (x);				      \
-	 else								      \
-	   __ll = __bswap_non_constant_64 (x);				      \
-	 __ll; }))
+# if (__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 3))
+#  define __bswap_64(x) \
+     __builtin_bswap64 ((unsigned long long int) (x))
+# else
+#  define __bswap_64(x) \
+     (__extension__							      \
+      ({ __extension__ unsigned long long int __ll =			      \
+	   __bswap_constant_64 (x);					      \
+	 __ll; }))
+# endif
 #endif
 
 #endif /* _BITS_BYTESWAP_H */
-- 
2.39.5
